rust_snuba/src/processors/release_health_metrics.rs

```diff
@@ -0,0 +1,761 @@
+/// tag keys are indexed so they should be integers. But we have to deal with it.
+fn generate_timeseries_id(
+    org_id: u64,
+    project_id: u64,
+    metric_id: u64,
+    tags: &BTreeMap<String, u64>,
+) -> u32 {
+    let mut adler = Adler32::new();
+
+    adler.write_slice(&org_id.to_le_bytes());
+    adler.write_slice(&project_id.to_le_bytes());
+    adler.write_slice(&metric_id.to_le_bytes());
+
+    for (key, value) in tags {
+        adler.write_slice(key.as_bytes());
+        adler.write_slice(&value.to_le_bytes());
+    }
+
+    adler.checksum()
+}
+
+#[derive(Debug, Deserialize)]
+struct FromMetricsMessage {
+    use_case_id: String,
+    org_id: u64,
+    project_id: u64,
+    metric_id: u64,
+    #[serde(rename = "type")]
+    r#type: String,
+    timestamp: f64,
+    sentry_received_timestamp: Option<f64>,
+    tags: BTreeMap<String, u64>,
+    value: MetricValue,
+    retention_days: u16,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum MetricValue {
+    Counter(f64),
+    SetOrDistribution(Vec<u64>),
+    DistributionFloat(Vec<f64>),
+}
+
+/// The raw row that is written to clickhouse.
+#[derive(Debug, Serialize, Default)]
+struct MetricsRawRow {
+    use_case_id: String,
+    org_id: u64,
+    project_id: u64,
+    metric_id: u64,
+    timestamp: u32,
+    retention_days: u16,
+    #[serde(rename = "tags.key")]
+    tags_key: Vec<u64>,
+    #[serde(default, rename = "tags.value")]
+    tags_value: Vec<u64>,
+    metric_type: String,
+    set_values: Option<Vec<u64>>,
+    count_value: Option<f64>,
+    distribution_values: Option<Vec<f64>>,
+    materialization_version: u8,
+    timeseries_id: u32,
+    partition: u16,
+    offset: u64,
+}
+
+/// Parse is the trait which should be implemented for all metric types.
+/// It is used to parse the incoming message into the appropriate raw row.
+/// Item represents the row into which the message should be parsed.
+trait Parse {
+    fn parse(
+        from: FromMetricsMessage,
+        meta: KafkaMessageMetadata,
+        config: &ProcessorConfig,
+    ) -> anyhow::Result<Option<MetricsRawRow>>;
+}
+
+impl Parse for MetricsRawRow {
+    fn parse(
+        from: FromMetricsMessage,
+        meta: KafkaMessageMetadata,
+        config: &ProcessorConfig,
+    ) -> anyhow::Result<Option<MetricsRawRow>> {
+        let timeseries_id =
+            generate_timeseries_id(from.org_id, from.project_id, from.metric_id, &from.tags);
+        let (tag_keys, tag_values): (Vec<_>, Vec<_>) = from.tags.into_iter().unzip();
+        let retention_days = enforce_retention(Some(from.retention_days), &config.env_config);
+
+        // The way the _should_process functionality works in the Python
+        // processor today is by matching against the InputType Enum
+        // Maybe we should look into porting that Enum into Rust?
+        match from.r#type.as_str() {
+            "s" => Ok(Some(MetricsRawRow {
+                use_case_id: from.use_case_id,
```

Comment:
A lot of the repetition between fields which are common between different message types can be abstracted away into a common struct. See [this](https://github.com/getsentry/snuba/blob/master/rust_snuba/src/processors/generic_metrics.rs#L82-L107) for example. You can do something similar to make the code more concise.

Reply:
Done, didn't use a common struct but did something similar,  please let me know if it looks good